components/brave_rewards/browser/rewards_service_impl.cc

```diff
@@ -1190,4 +1207,75 @@ double RewardsServiceImpl::GetContributionAmount() {
   return ledger_->GetContributionAmount();
 }
 
+void RewardsServiceImpl::FetchFavIcon(const std::string& url, const std::string& favicon_key) {
+  GURL parsedUrl(url);
+
+  if (!parsedUrl.is_valid()) {
+    return;
+  }
+
+  std::vector<std::string>::iterator it;
+  it = find (current_media_fetchers_.begin(), current_media_fetchers_.end(), url);
+  if (it != current_media_fetchers_.end()) {
+    VLOG(1) << "Already fetching favicon: " << url;
+    return;
+  }
+
+  current_media_fetchers_.emplace_back(url);
+
+  BitmapFetcherService* image_service =
+      BitmapFetcherServiceFactory::GetForBrowserContext(profile_);
+  if (image_service) {
+    net::NetworkTrafficAnnotationTag traffic_annotation =
+      net::DefineNetworkTrafficAnnotation("brave_rewards_favicon_fetcher", R"(
+        semantics {
+          sender:
+            "Brave Rewards Media Fetcher"
+          description:
+            "Fetches favicon for media publishers in Rewards."
+          trigger:
+            "User visits a media publisher content."
+          data: "Favicon for media publisher."
+          destination: WEBSITE
+        }
+        policy {
+          cookies_allowed: NO
+          setting:
+            "This feature cannot be disabled by settings."
+          policy_exception_justification:
+            "Not implemented."
+        })");
+    request_ids_.push_back(image_service->RequestImage(
+          parsedUrl,
+          // Image Service takes ownership of the observer
+          new RewardsFetcherServiceObserver(
+              favicon_key,
+              parsedUrl,
+              base::Bind(&RewardsServiceImpl::OnFetchFavIconCompleted, base::Unretained(this))),
+          traffic_annotation));
+  }
+}
+
+void RewardsServiceImpl::OnFetchFavIconCompleted(const std::string& favicon_key,
+                                            const GURL& url,
+                                            const BitmapFetcherService::RequestId& request_id,
+                                            const SkBitmap& image) {
+  GURL publisher_url(favicon_key);
+  gfx::Image gfx_image = gfx::Image::CreateFrom1xBitmap(image);
+  favicon::FaviconService* favicon_service =
+          FaviconServiceFactory::GetForProfile(profile_, ServiceAccessType::EXPLICIT_ACCESS);
+  favicon_service->SetOnDemandFavicons(
+      publisher_url,
+      url, favicon_base::IconType::kFavicon, gfx_image,
+      base::BindOnce(&RewardsServiceImpl::OnSetOnDemandFaviconComplete, AsWeakPtr()));
+
+  // TODO remove request_id from the array
```

Comment:
Yep was just about to add a comment here and then seen the comment :)
Please remove the ID from request_ids_ otherwise there's a slow memory leak that grows forever.

Reply:
forgot about this one, will fix

Reply:
done